### PR TITLE
community: Load YouTube transcripts (captions) as fixed-duration chunks with start times

### PR DESCRIPTION
**Pull request by lsloan**
- **Description:** Add a new format, `CHUNKS`, to `langchain_community.document_loaders.youtube.YoutubeLoader` which creates multiple `Document` objects from YouTube video transcripts (captions), each of a fixed duration.  The metadata of each chunk `Document` includes the start time of each one and a URL to that time in the video on the YouTube website.
  
  I had implemented this for UMich (@umich-its-ai) in a local module, but it makes sense to contribute this to LangChain community for all to benefit and to simplify maintenance.

- **Issue:** N/A
- **Dependencies:** N/A
- **Twitter:** lsloan_umich
- **Mastodon:** [lsloan@mastodon.social](https://mastodon.social/@lsloan)

With regards to **tests and documentation**, most existing features of the `YoutubeLoader` class are not tested.  Only the `YoutubeLoader.extract_video_id()` static method had a test.  However, while I was waiting for this PR to be reviewed and merged, I had time to add a test for the chunking feature I've proposed in this PR. 

I have added an example of using chunking to the `docs/docs/integrations/document_loaders/youtube_transcript.ipynb` notebook.

### CODE REVIEW

**Comment by baskaryan** [comment]
@lsloan feel free to tag when pr is ready for review

**Comment by lsloan** [comment]
@baskaryan, I've been having trouble getting Poetry installed and running the initial `make test` in `libs/community`.  When I finally got Poetry running, the test gave the error: `pytest: error: unrecognized arguments: --snapshot-warn-unused`.

Should I remove that argument from `pyproject.toml`?

**Comment by lsloan** [comment]
I've tried variations of the env. setup, to no avail thus far.  My latest, simplest attempt which illustrates a problem I saw yesterday…

Foregoing the recommendation of using `pipx`, my steps are…

1. `brew upgrade python@3.9`
   * Based on LangChain quick start docs which say to use a Conda env. based on 3.9. 
2. `cd ~/Projects/maizey/langchain/`
3. `python3.9 -mvenv .venv`
4. `. .venv/bin/activate`
5. `cd libs/community`
6. `pip install poetry`
7. `poetry install --with lint,typing,test,test_integration`
    * ```txt
        - Installing torch (2.3.0): Failed
      
        RuntimeError
      
        Unable to find installation candidates for torch (2.3.0)
      
        at ~/Projects/maizey/langchain/.venv/lib/python3.9/site-packages/poetry/installation/chooser.py:74 in choose_for
             70│ 
             71│             links.append(link)
             72│ 
             73│         if not links:
          →  74│             raise RuntimeError(f"Unable to find installation candidates for {package}")
             75│ 
      ```
      
When I first saw this error yesterday, I thought that maybe I was getting too new a version of Poetry.  I tried to use an earlier version of it, 1.7.1, as I described above, but that gave errors about pytest unknown arguments.

I would be grateful for advice.  I don't normally use conda.  I tried it years ago and didn't like it much.  I've been working well without it.  That's the only difference I see between my environment and what the LangChain quick start documentation recommends.

**Comment by lsloan** [comment]
@baskaryan, thanks for the advice.  I made a little more progress with that.  However, when I run `make test`, I get this error…

```txt
poetry run pytest tests/unit_tests/
Traceback (most recent call last):
⋮
[some traceback lines omitted]
⋮
  File "/Users/lsloan/Projects/maizey/langchain/.venv/lib/python3.9/site-packages/cleo/_utils.py", line 8, in <module>
    from rapidfuzz.distance import Levenshtein
ModuleNotFoundError: No module named 'rapidfuzz'
make: *** [test] Error 1
```

This is interesting because when I ran `poetry install --with lint,typing,test`, the very first line of output was `  - Removing rapidfuzz (3.6.2)`.  Sure enough, when I check the installation with `pip freeze`, `rapidfuzz` is not in the list.  I wonder why this process removes that module if it's needed later.

**Comment by lsloan** [comment]
@baskaryan, thank you for your advice!

This PR is ready for review. 🤞

**Review by baskaryan — APPROVED** [review]

File: libs/community/langchain_community/document_loaders/youtube.py [hunk 1]
```diff
@@ -163,7 +163,7 @@ def __init__(
     ):
         """Initialize with YouTube video ID."""
         self.video_id = video_id
-        self.metadata = {"source": video_id}
+        self._metadata = {"source": video_id}
```
Comment (lsloan) [inline libs/community/langchain_community/document_loaders/youtube.py:166]: @baskaryan, thanks.  That seems like a reasonable change.

**Comment by lsloan** [comment]
@baskaryan, also, thanks for reviewing and merging this PR.  🎉 :shipit: 

I have some additional refactoring and possibly more unit tests for the YouTube transcript loader.  I'll open a new PR for that when it's ready.

**Pull request merged**